internal/funcs/generated/gen/gen.go

```diff
@@ -0,0 +1,286 @@
+}
+
+// signaturesFromTerraform gets the function signatures for the specified
+// Terraform version.
+func signaturesFromTerraform(ctx context.Context) (*tfjson.MetadataFunctions, error) {
+	// find or install Terraform
+	log.Println("ensuring terraform is installed")
+	installDir, err := ioutil.TempDir("", "hcinstall")
+	if err != nil {
+		return nil, err
+	}
+	defer os.RemoveAll(installDir)
+	i := hcinstall.NewInstaller()
+	execPath, err := i.Ensure(ctx, []src.Source{
+		&releases.ExactVersion{
+			Product:    product.Terraform,
+			InstallDir: installDir,
+			Version:    version.Must(version.NewVersion(terraformVersion)),
+		},
+	})
+	if err != nil {
+		return nil, err
+	}
+	defer i.Remove(ctx)
+
+	// log version
+	tf, err := tfexec.NewTerraform(installDir, execPath)
+	if err != nil {
+		return nil, err
+	}
+	coreVersion, _, err := tf.Version(ctx, true)
+	if err != nil {
+		return nil, err
+	}
+	log.Printf("using Terraform %s (%s)", coreVersion, execPath)
+
+	return tf.MetadataFunctions(ctx)
+}
+
+// signatureHash calculates the SHA256 checksum for the given function
+// signatures.
+func signatureHash(functions *tfjson.MetadataFunctions) (string, error) {
+	var rawJson bytes.Buffer
+	err := json.NewEncoder(&rawJson).Encode(functions)
+	if err != nil {
+		return "", fmt.Errorf("failed to encode functions: %w", err)
+	}
+
+	hash := sha256.Sum256(rawJson.Bytes())
+	return fmt.Sprintf("%x", hash), nil
+}
+
+// writeFunctions generates the Go code for the function signatures,
+// formats it and writes it to a new file.
+func writeFunctions(filename string, functions *tfjson.MetadataFunctions) {
+	outputTpl := `// Code generated by "gen"; DO NOT EDIT.
+package funcs
+
+import (
+	"github.com/hashicorp/hcl-lang/schema"
+	"github.com/zclconf/go-cty/cty"
+	"github.com/zclconf/go-cty/cty/function"
+)
+func {{ .FunctionName }}() map[string]schema.FunctionSignature {
+	return map[string]schema.FunctionSignature{
+{{- range $funcName, $func := .Signatures }}
+		{{- $varpar := $func.VariadicParameter }}
+		{{- $params := $func.Parameters }}
+		"{{ $funcName }}": {
+			{{- with $params }}
+			Params: []function.Parameter{
+				{{- range $key, $value := $params }}
+				{{- $desc := $value.Description }}
+				{
+					Name:        "{{ $value.Name }}",
+					{{- with $desc }}
+					Description: "{{ escapeDescription . }}",{{- end }}
+					Type:        {{ $value.Type.GoString }},
+				},
+				{{- end }}
+			},{{- end }}
+			{{- with $varpar }}
+			VarParam:    &function.Parameter{
+				Name:             "{{ .Name }}",
+				Description:      "{{ .Description }}",
+				Type:             {{ .Type.GoString }},
+			},{{- end }}
+			ReturnType: {{ $func.ReturnType.GoString }},
+			Description: "{{ escapeDescription $func.Description }}",
+		},	
+{{- end }}
+	}
+}
+`
+
+	tpl, err := template.New("output").Funcs(template.FuncMap{
+		"escapeDescription": escapeDescription,
+	}).Parse(outputTpl)
+	if err != nil {
+		log.Fatal(err)
```

Comment:
Nitpick: I know this is a script which will be executed by a maintainer, but it's still a good practice for any downstream functions to _return_ error and only the top level to actually deal with surfacing it somehow.

This enables testability and reusability and helps readability (it's easier to understand the flow of the program as a whole), even though admittedly we probably won't be testing or reusing this code any time soon.

Reply:
Oh, missed updating the error handling for the two templating functions. Thanks for pointing it out.